packages/trie/src/db.ts

```diff
@@ -55,23 +106,75 @@ export class DB {
    * @param value The value to be stored
    */
   async put(key: Buffer, val: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
+          type: 'put',
+          key: key,
+          value: oldValue,
+        })
+      }
+    }
     await this._leveldb.put(key, val, ENCODING_OPTS)
+    // In CP mode add del to the revert ops stack
+    if (this.isCheckpoint) {
+      revertOps.push({
+        type: 'del',
+        key: key,
+      })
+      for (const revertOp of revertOps) {
+        this.addCPRevertOperation(revertOp)
+      }
+    }
   }
 
   /**
    * Removes a raw value in the underlying leveldb.
    * @param keys
    */
   async del(key: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
```

Comment:
Since we only put one operation in the `revertOps` array, can't we immediately call `addCPRevertOperation`?

Reply:
This is for the case that `leveldb.del` might throw for some reason, then the revert operation shouldn't be added to the revert stack.